components/script/dom/permissions.rs

```diff
@@ -0,0 +1,245 @@
+use dom::bindings::codegen::Bindings::PermissionStatusBinding::{PermissionDescriptor, PermissionName};
+use dom::bindings::codegen::Bindings::PermissionsBinding::{self, PermissionsMethods};
+use dom::bindings::error::Error;
+use dom::bindings::js::Root;
+use dom::bindings::reflector::{DomObject, Reflector, reflect_dom_object};
+use dom::bluetoothpermissionresult::BluetoothPermissionResult;
+use dom::globalscope::GlobalScope;
+use dom::permissionstatus::{PermissionDescriptorType, PermissionStatus};
+use dom::promise::Promise;
+use js::conversions::{ConversionResult, FromJSValConvertible};
+use js::jsapi::{JSContext, JSObject};
+use js::jsval::{ObjectValue, UndefinedValue};
+use std::rc::Rc;
+
+const ROOT_DESC_CONVERSION_ERROR: &'static str = "Can't convert to an IDL value of type PermissionDescriptor";
+const BT_DESC_CONVERSION_ERROR: &'static str = "Can't convert to an IDL value of type BluetoothPermissionDescriptor";
+
+// https://w3c.github.io/permissions/#permissions
+#[dom_struct]
+pub struct Permissions {
+    reflector_: Reflector,
+}
+
+impl Permissions {
+    pub fn new_inherited() -> Permissions {
+        Permissions {
+            reflector_: Reflector::new(),
+        }
+    }
+
+    pub fn new(global: &GlobalScope) -> Root<Permissions> {
+        reflect_dom_object(box Permissions::new_inherited(),
+                           global,
+                           PermissionsBinding::Wrap)
+    }
+}
+
+#[allow(unsafe_code)]
+fn create_root_descriptor(cx: *mut JSContext,
+                          permission_descriptor_obj: *mut JSObject)
+                          -> Result<PermissionDescriptor, String> {
+    rooted!(in(cx) let mut property = UndefinedValue());
+    property.handle_mut().set(ObjectValue(permission_descriptor_obj));
+    unsafe {
+        match PermissionDescriptor::from_jsval(cx, property.handle(), ()) {
+            Ok(ConversionResult::Success(descriptor)) => Ok(descriptor),
+            Ok(ConversionResult::Failure(message)) => Err(String::from(message)),
+            Err(_) => Err(String::from(ROOT_DESC_CONVERSION_ERROR)),
+        }
+    }
+}
+
+#[allow(unsafe_code)]
+fn create_bluetooth_descriptor(cx: *mut JSContext,
+                               permission_descriptor_obj: *mut JSObject)
+                               -> Result<BluetoothPermissionDescriptor, String> {
+    rooted!(in(cx) let mut property = UndefinedValue());
+    let jsval = ObjectValue(permission_descriptor_obj);
+    property.handle_mut().set(jsval);
+    unsafe {
+        match BluetoothPermissionDescriptor::from_jsval(cx, property.handle(), ()) {
+            Ok(ConversionResult::Success(descriptor)) => Ok(descriptor),
+            Ok(ConversionResult::Failure(message)) => Err(String::from(message)),
+            Err(_) => Err(String::from(BT_DESC_CONVERSION_ERROR)),
+        }
+    }
+}
+
+// https://w3c.github.io/permissions/#dom-permissions-query
+fn sync_default_permission_query_call(global: &GlobalScope,
```

Comment:
The same applies here, i think we can do these `sync_*_permission_query_call`s generic with enum or trait.